### ISSUE TITLE
VersionedApiExplorer breaks something in OpenIddict or ASOS

### ISSUE THREAD

**Issue by xperiandri** [Microsoft/aspnet-api-versioning#134]
After I added `services.AddMvcCore().AddVersionedApiExplorer();` the request to `http://localhost:1966/connect/token` ends with 404

@PinpointTownes, how can we investigate this?

    Microsoft.AspNetCore.Hosting.Internal.WebHost: Information: Request starting HTTP/1.1 POST http://localhost:1966/connect/token application/x-www-form-urlencoded 158
    The thread 0xd54 has exited with code 0 (0x0).
    Microsoft.EntityFrameworkCore.Storage.IRelationalCommandBuilderFactory: Information: Executed DbCommand (34ms) [Parameters=[@__get_Item_0='?'], CommandType='Text', CommandTimeout='30']
    SELECT TOP(1) [e].[Id], [e].[AccessFailedCount], [e].[ConcurrencyStamp], [e].[Email], [e].[EmailConfirmed], [e].[IsDriver], [e].[LockoutEnabled], [e].[LockoutEnd], [e].[NormalizedEmail], [e].[NormalizedUserName], [e].[PasswordHash], [e].[PhoneNumber], [e].[PhoneNumberConfirmed], [e].[PictureUrl], [e].[QuickBloxId], [e].[QuickBloxPassword], [e].[SecurityStamp], [e].[TwilioSid], [e].[TwoFactorEnabled], [e].[UserName]
    FROM [Users] AS [e]
    WHERE [e].[IsDriver] IN (0, 1) AND ([e].[Id] = @__get_Item_0)
    The thread 0x2130 has exited with code 0 (0x0).
    Microsoft.EntityFrameworkCore.Storage.IRelationalCommandBuilderFactory: Information: Executed DbCommand (11ms) [Parameters=[@p0='?' (Size = 4000)], CommandType='Text', CommandTimeout='30']
    
    SELECT * FROM BhutanDrivers
    WHERE [UserId] = @p0
    The thread 0x1bd4 has exited with code 0 (0x0).
    Microsoft.EntityFrameworkCore.Storage.IRelationalCommandBuilderFactory: Information: Executed DbCommand (25ms) [Parameters=[@__userId_0='?'], CommandType='Text', CommandTimeout='30']
    SELECT [role].[Name]
    FROM [RolesOfUsers] AS [userRole]
    INNER JOIN [UserRoles] AS [role] ON [userRole].[RoleId] = [role].[Id]
    WHERE [userRole].[UserId] = @__userId_0
    The thread 0x2384 has exited with code 0 (0x0).
    Microsoft.EntityFrameworkCore.Storage.IRelationalCommandBuilderFactory: Information: Executed DbCommand (6ms) [Parameters=[@__user_Id_0='?'], CommandType='Text', CommandTimeout='30']
    SELECT [uc].[Id], [uc].[ClaimType], [uc].[ClaimValue], [uc].[UserId]
    FROM [UserClaims] AS [uc]
    WHERE [uc].[UserId] = @__user_Id_0
    Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationMiddleware: Information: HttpContext.User merged via AutomaticAuthentication from authenticationScheme: Identity.Application.
    AspNet.Security.OpenIdConnect.Server.OpenIdConnectServerMiddleware: Information: The token request was successfully extracted from the HTTP request: {
      "username": "andriicsharp",
      "password": "[removed for security reasons]",
      "client_id": "postman",
      "grant_type": "password",
      "scope": "openid email phone profile qb tw offline",
      "device_id": "Lumia929"
    }.
    Microsoft.EntityFrameworkCore.Storage.IRelationalCommandBuilderFactory: Information: Executed DbCommand (3ms) [Parameters=[@__identifier_0='?' (Size = 450)], CommandType='Text', CommandTimeout='30']
    SELECT TOP(2) [application].[Id], [application].[ClientId], [application].[ClientSecret], [application].[DisplayName], [application].[LogoutRedirectUri], [application].[PushService], [application].[RedirectUri], [application].[Type]
    FROM [OpenIddictApplications] AS [application]
    WHERE [application].[ClientId] = @__identifier_0
    OpenIddict.OpenIddictProvider: Information: The token request validation process was not fully validated because the client 'postman' was a public application.
    AspNet.Security.OpenIdConnect.Server.OpenIdConnectServerMiddleware: Information: The token request was successfully validated.
    Microsoft.AspNetCore.Hosting.Internal.WebHost: Information: Request finished in 808.0164ms 404

**Comment by xperiandri** [comment]
``` C#
[ApiVersionNeutral]
[Route("[controller]")]
public class AuthorizeController : Controller
{
    [HttpPost("~/connect/token")]
    [Produces("application/json")]
    [ProducesResponseType(typeof(void), (int)HttpStatusCode.OK)]
    [ProducesResponseType(typeof(OpenIdConnectResponse), (int)HttpStatusCode.BadRequest)]
    public async Task<IActionResult> Exchange([FromServices] OpenIdConnectRequest request)
```

**Comment by PinpointTownes** [comment]
> @PinpointTownes, how can we investigate this?

It's likely not an OpenIddict issue. To confirm it, change the default log level to `Debug` or `Verbose`: if you see the `AspNet.Security.OpenIdConnect.Server.OpenIdConnectServerMiddleware: Debug: The default token request handling was skipped from user code.` message (that indicates that ASOS was asked to invoke the next middleware in the pipeline), you can be sure it's not caused by OpenIddict.

I guess adding the API versioning stack changes how MVC handles the `/connect/token` route and indirectly prevents your `Exchange` action from being properly invoked.

**Comment by xperiandri** [comment]
@PinpointTownes, this way?

``` JSON
  "Logging": {
    "Console": {
      "IncludeScopes": true,
      "LogLevel": {
        "Default": "Debug",
        "System": "Error",
        "Microsoft": "Error"
      }
    }
  },
```

**Comment by commonsensesoftware** [comment]
The API explorer has no influence on routing. I don't see how (yet) that would adversely affect anything. @xperiandri is this on the 1.1 RC? If so, this issue is probably related to #133. There's a regression that could cause this. I'm working on the fix as fast as possible. If things were working up until Beta 2, then I suspect that's the culprit.

**Comment by PinpointTownes** [comment]
@xperiandri you should consider using the Debug logger. VS's debug output window is way better for logging than the default console :smile:

**Comment by xperiandri** [comment]
@PinpointTownes, could you drop a sample or a link to some info?

**Comment by PinpointTownes** [comment]
Google is your friend. But since I'm a cool guy... the documentation is here: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging.

**Comment by xperiandri** [comment]
@PinpointTownes, thanks.
So I had to put `LogLevel` instead of `Console` in my config?

@commonsensesoftware, looks like that.

**Comment by xperiandri** [comment]
@commonsensesoftware, after downgrading to `Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer` `1.0.0-beta1` it works fine

**Comment by commonsensesoftware** [comment]
@xperiandri  What about `Microsoft.AspNetCore.Mvc.Versioning` itself? Is that `1.1.0-beta1`, `1.1.0-beta2`, `1.1.0-rc1`? If you're only taking the dependency on `Microsoft.AspNetCore.Mvc.Versioning`, then it will transitively match `1.1.0-beta2`. That's generally _good_ news because this and several other issues appear to be all caused by the same regression.

I'm trying to get a fix out ASAP. I have things fixed, but I'm expanding my acceptance test coverage to add more of the issues reported to make sure things are really fixed. I don't want to just arbitrarily throw something out there though.

**Comment by xperiandri** [comment]
Yes, transitively `1.1.0-beta2`
Agree, you are doing right.

**Comment by commonsensesoftware** [comment]
[RC2](https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc.Versioning/1.1.0-rc2) has been officially published. Let me know if you run into any other issues.

**Comment by xperiandri** [comment]
Doesn't work. Still getting 404

**Comment by commonsensesoftware** [comment]
A repro project would be useful; however, before I ask you to do a lot of work, can you share you `Startup.cs` configuration? Thanks.

**Comment by commonsensesoftware** [comment]
I looked at your configuration and you're missing the call to `UseApiVersioning`.  This is a behavioral, break-breaking change in ASP.NET Core in the RC. It's called out as loudly as I could in the release notes. It was the only option in ASP.NET Core's current design to enable a comprehensive set of out-of-the-box routing behaviors. I've also updated the wiki to reflect this.

Your `Configure` method must now have:

```c#
// ...
app.UseMvc();
app.UseApiVersioning();
// ...
```

Try that and see if things work.
